src/kernel.cu

```diff
@@ -0,0 +1,131 @@
+#include <cstdlib>
+#include <stdio.h>
+#include <float.h>
+
+#include "point.h"
+
+//Testing for CUDA errors
+inline cudaError_t checkCuda(cudaError_t result) {
+if (result != cudaSuccess) {
+fprintf(stderr, "CUDA Runtime Error: %s\n", cudaGetErrorString(result));
+}
+return result;
+}
+
+//Clustering Calculation
+__global__ 
+void my_Kernel_1(Point* points_d, int points_size, Point* c_d, int cluster_id) {
+    
+    //Get the index for the current point to work with
+    int Col = threadIdx.x + blockIdx.x * blockDim.x;
+    int Row = threadIdx.y + blockIdx.y * blockDim.y;
+    int index = Col + (Row * blockDim.x *gridDim.x); 
+    
+    if(index < points_size) {
+        Point p = points_d[index];
+        double dist = ((c_d->x - p.x) * (c_d->x - p.x) + (c_d->y - p.y) * (c_d->y - p.y) + (c_d->z - p.z) * (c_d->z - p.z));// point_distance(c, *p);
+        if (dist < p.min_dist) {
+            p.min_dist = dist;
+            p.cluster = cluster_id;
+        }
+        
+        points_d[index] = p;
+    }
+}
+
+__global__
+void my_Kernel_2(double max, Point* points_d, int points_size, int* n_points_d, double* sum_x_d, double* sum_y_d, double* sum_z_d) {
+    //Get the index of the point to work with
+    int Col = threadIdx.x + blockIdx.x * blockDim.x;
+    int Row = threadIdx.y + blockIdx.y * blockDim.y;
+    int index = Col + (Row * blockDim.x * gridDim.x); 
+    
+    if(index < points_size) {
+        Point p = points_d[index];
+        n_points_d[p.cluster] += 1;
```

Comment:
I think there's a race condition here, with all the threads trying to update the array at the same time